lib/auth/keystore/manager.go

```diff
@@ -120,24 +137,33 @@ func NewManager(ctx context.Context, cfg Config) (*Manager, error) {
 		return nil, trace.Wrap(err)
 	}
 
-	logger := cfg.Logger
-	if logger == nil {
-		logger = logrus.StandardLogger()
-	}
+	softwareBackend := newSoftwareKeyStore(&cfg.Software, cfg.Logger)
 
 	if (cfg.PKCS11 != PKCS11Config{}) {
-		backend, err := newPKCS11KeyStore(&cfg.PKCS11, logger)
-		return &Manager{backend: backend}, trace.Wrap(err)
+		pkcs11Backend, err := newPKCS11KeyStore(&cfg.PKCS11, cfg.Logger)
+		return &Manager{
+			backendForNewKeys:     pkcs11Backend,
+			usableSigningBackends: []backend{pkcs11Backend, softwareBackend},
+		}, trace.Wrap(err)
 	}
 	if (cfg.GCPKMS != GCPKMSConfig{}) {
-		backend, err := newGCPKMSKeyStore(ctx, &cfg.GCPKMS, logger)
-		return &Manager{backend: backend}, trace.Wrap(err)
+		gcpBackend, err := newGCPKMSKeyStore(ctx, &cfg.GCPKMS, cfg.Logger)
+		return &Manager{
+			backendForNewKeys:     gcpBackend,
+			usableSigningBackends: []backend{gcpBackend, softwareBackend},
+		}, trace.Wrap(err)
 	}
 	if (cfg.AWSKMS != AWSKMSConfig{}) {
-		backend, err := newAWSKMSKeystore(ctx, &cfg.AWSKMS, logger)
-		return &Manager{backend: backend}, trace.Wrap(err)
+		awsBackend, err := newAWSKMSKeystore(ctx, &cfg.AWSKMS, cfg.Logger)
+		return &Manager{
+			backendForNewKeys:     awsBackend,
+			usableSigningBackends: []backend{awsBackend, softwareBackend},
+		}, trace.Wrap(err)
 	}
-	return &Manager{backend: newSoftwareKeyStore(&cfg.Software, logger)}, nil
+	return &Manager{
+		backendForNewKeys:     softwareBackend,
+		usableSigningBackends: []backend{softwareBackend},
+	}, nil
```

Comment:
This is the meat of the actual change. Instead of only having a single keystore backend for everything, the keystore manager has one preferred backend for any new keys it will generate (this is only used to generate CA keys) and a list of backends it can use to sign stuff, in preference order. I always include the software keystore as the last element in the list, so the auth can always sign certs if there are any software keys in the CA.